src/glang/codegen/floats.py

```diff
@@ -0,0 +1,124 @@
+from decimal import Decimal
+
+from .user_facing_errors import (
+    InvalidFloatLiteralPrecision,
+    InvalidFloatLiteralTooLarge,
+)
+
+EXPONENT_BIT_TABLE = {16: 5, 32: 8, 64: 11, 128: 15}
+FRACTION_BIT_TABLE = {16: 10, 32: 23, 64: 52, 128: 112}
+
+
+def fraction_to_binary_list(
+    number: Decimal, precision_in_bits: int, list_length: int
+) -> list[bool]:
+    assert number <= 1
+    without_decimal_place = int(number * 2**precision_in_bits)
+
+    result = [False] * (list_length - precision_in_bits)
+    for _ in range(precision_in_bits):
+        result.append((without_decimal_place & 1) == 1)
+        without_decimal_place >>= 1
+
+    assert without_decimal_place == 0
+    return result
+
+
+def float_literal_to_exact_hex(input_float_str: str, float_size_in_bits: int) -> str:
+    float_str = input_float_str.lower()
+
+    # We support both exponential and decimal notation
+    # First simplify the algorithm by inserting the implicit exponent
+    if "e" not in float_str:
+        float_str += "e0"
+
+    # Now trim the sign (if any)
+    # IEEE floats have symmetric precision so both cases are identical
+    is_negative = float_str[0] == "-"
+    if float_str[0] == "-" or float_str[0] == "+":
+        float_str = float_str[1:]
+
+    mantissa_str_10, exponent_str_10 = float_str.split("e")
+
+    # The exponent is an integer, python can give us an exact representation
+    exponent_10 = int(exponent_str_10)
+
+    # The mantissa may contain a decimal place, adjust the exponent until it is an integer
+    if "." in mantissa_str_10:
+        decimal_point_index = mantissa_str_10.index(".")
+
+        mantissa_str_10 = mantissa_str_10.replace(".", "")
+        exponent_10 += decimal_point_index - len(mantissa_str_10)
+
+    mantissa_10 = int(mantissa_str_10)
+
+    if mantissa_10 == 0:
+        # Note: I'm discarding '-0.0' here... I'm not sure if its worth propagating
```

Comment:
Do we parse the leading `-` in integer literals?

Reply:
Sometimes? We do in generics but not in expressions. Floats should behave in exactly the same way but since they're not allowed in generics ur right, `-0.0` won't actually be parsable even if I implemented it.

Reply:
Guhh, that's not very nice. Compile-time expressions would make this more consistent :)